docs/graph/code/hld/hld_2.cpp

```diff
@@ -0,0 +1,223 @@
+      // printf("st[%d] <- idx[%d] = %d\n", p, gl, idx[gl]->val);
+      return (void)(st[p] = idx[gl = gr]->val);
+    }
+    Build(LS, gl, MID);
+    Build(RS, MID + 1, gr);
+    Pushup(p);
+  }
+
+  void Modify(int l, int r, int val, int p = 1, int gl = 1, int gr = N) {
+    if (l <= gl && gr <= r)
+      return (void)(st[p] = (st[p] + SIZ * val) % MOD,
+                    lt[p] = (lt[p] + val) % MOD);
+    if (lt[p]) Pushdown(p, gl, gr);
+    if (l <= MID) Modify(l, r, val, LS, gl, MID);
+    if (r >= MID + 1) Modify(l, r, val, RS, MID + 1, gr);
+    Pushup(p);
+  }
+
+  ll Query(int l, int r, int p = 1, int gl = 1, int gr = N) {
+    if (l <= gl && gr <= r) return st[p];
+    if (lt[p]) Pushdown(p, gl, gr);
+    return ((l <= MID ? Query(l, r, LS, gl, MID) : 0) +
+            (r >= MID + 1 ? Query(l, r, RS, MID + 1, gr) : 0)) %
+           MOD;
+  }
+
+  void Desc(void) {
+    int cur(0);
+    for (int i = 1; i <= 1 << 4; i <<= 1)
+      for (int j = 1; j <= i; ++j)
+        printf("%lld%c", st[++cur], j == i ? '\n' : ' ');
+  }
+} st;
+
+#define top(p) (p ? p->top : npt)
+#define dfn(p) (p ? p->dfn : 0)
+
+void ModifyTree(Node* from, Node* to, int val) {
+  while (top(from) != top(to)) {
+    if (dep(top(from)) < dep(top(to))) swap(from, to);
+    st.Modify(dfn(from->top), dfn(from), val);
+    from = from ? from->top->fa : npt;
+  }
+  if (dep(from) < dep(to)) swap(from, to);
+  st.Modify(dfn(to), dfn(from), val);
+}
+
+ll QueryTree(Node* from, Node* to) {
+  ll ret(0);
+  while (top(from) != top(to)) {
+    if (dep(top(from)) < dep(top(to))) swap(from, to);
+    ret = (ret + st.Query(dfn(from->top), dfn(from))) % MOD;
+    from = from ? from->top->fa : npt;
+  }
+  if (dep(from) < dep(to)) swap(from, to);
+  ret = (ret + st.Query(dfn(to), dfn(from))) % MOD;
+  return ret;
+}
+
+int main() {
+  N = read(), M = read(), R = read(), MOD = read();
+  for (int i = 1; i <= N; ++i) tr[i] = new Node(read());
+  for (int i = 1; i <= N - 1; ++i) {
+    int f = read(), t = read();
+    tr[f]->head = new Edge(tr[f]->head, tr[t]);
+    tr[t]->head = new Edge(tr[t]->head, tr[f]);
+  }
+  dfs_pretreat(tr[R], npt);
+  dfs_build(tr[R], tr[R]);
+  st.Build();
+  while (M--) {
+    int opt = read();
+    switch (opt) {
+      case 1: {
+        int f = read(), t = read(), val = read() % MOD;
+        ModifyTree(tr[f], tr[t], val);
+        break;
+      }
+      case 2: {
+        int f = read(), t = read();
+        printf("%lld\n", QueryTree(tr[f], tr[t]));
+        break;
+      }
+      case 3: {
+        int x = read(), val = read() % MOD;
+        st.Modify(tr[x]->dfn, tr[x]->dfn + tr[x]->siz - 1, val);
+        break;
+      }
+      case 4: {
+        int x = read();
+        printf("%lld\n", st.Query(tr[x]->dfn, tr[x]->dfn + tr[x]->siz - 1));
+        break;
+      }
+    }
+  }
+  // fprintf(stderr, "Time: %.6lf\n", (double)clock() / CLOCKS_PER_SEC);
+  return 0;
+}
+
+template <typename T>
+inline T read(void) {
+  T ret(0);
+  short flag(1);

```

Comment:
由于整数提升, 这里不如直接写成 `int` 🤔

```suggestion
  int flag(1);
```

Reply:
> 由于整数提升, 这里不如直接写成 `int` 🤔

查了一下，似乎这里确实没必要short，学到了学到了